File: max_finder_text.py
#!/usr/bin/env python3
"""
Text File Maximum Number Finder
Reads comma-separated integers from a text file and finds the maximum.

Supports:
- Single-core single-threaded mode
- Multi-core multi-threaded mode
- Real-time throughput and CPU monitoring
"""

import argparse
import csv
import json
import os
import sys
import threading
import time
from concurrent.futures import ThreadPoolExecutor, as_completed
from dataclasses import dataclass
from datetime import datetime
from typing import Optional, List, Dict

import psutil


@dataclass
class Stats:
    """Statistics for monitoring performance."""
    bytes_read: int = 0
    numbers_read: int = 0
    start_time: float = 0.0
    lock: threading.Lock = None
    
    def __post_init__(self):
        if self.lock is None:
            self.lock = threading.Lock()
    
    def add_progress(self, bytes_count: int, numbers_count: int):
        with self.lock:
            self.bytes_read += bytes_count
            self.numbers_read += numbers_count
    
    def get_throughput_mbps(self) -> float:
        elapsed = time.time() - self.start_time
        if elapsed > 0:
            return (self.bytes_read / (1024 * 1024)) / elapsed
        return 0.0


class IOWaitMonitor:
    """Monitor I/O wait time."""
    
    def __init__(self):
        self.wall_start = 0.0
        self.cpu_start = 0.0
        self.process = psutil.Process()
        
    def start(self):
        self.wall_start = time.time()
        cpu_times = self.process.cpu_times()
        self.cpu_start = cpu_times.user + cpu_times.system
        
    def get_stats(self) -> dict:
        wall_elapsed = time.time() - self.wall_start
        cpu_times = self.process.cpu_times()
        cpu_elapsed = (cpu_times.user + cpu_times.system) - self.cpu_start
        io_wait = max(0, wall_elapsed - cpu_elapsed)
        io_wait_pct = (io_wait / wall_elapsed * 100) if wall_elapsed > 0 else 0
        
        return {
            'wall_time': wall_elapsed,
            'cpu_time': cpu_elapsed,
            'io_wait_time': io_wait,
            'io_wait_pct': io_wait_pct,
            'cpu_busy_pct': min(100, (cpu_elapsed / wall_elapsed * 100) if wall_elapsed > 0 else 0),
            'user_time': cpu_times.user - self.cpu_start,
            'system_time': cpu_times.system
        }


class CPUMonitor:
    """Monitor CPU utilization per core."""
    
    def __init__(self, interval: float = 0.5):
        self.interval = interval
        self.running = False
        self.thread = None
        self.cpu_samples = []
        self.cpu_times_start = []
        self.cpu_times_end = []
        self.start_time = 0.0
        self.end_time = 0.0
        
    def start(self):
        self.running = True
        self.cpu_samples = []
        self.start_time = time.time()
        self.cpu_times_start = psutil.cpu_times(percpu=True)
        psutil.cpu_percent(percpu=True)
        self.thread = threading.Thread(target=self._monitor_loop, daemon=True)
        self.thread.start()
    
    def stop(self):
        self.running = False
        self.end_time = time.time()
        self.cpu_times_end = psutil.cpu_times(percpu=True)
        if self.thread:
            self.thread.join(timeout=2.0)
    
    def _monitor_loop(self):
        while self.running:
            cpu_percents = psutil.cpu_percent(percpu=True, interval=self.interval)
            self.cpu_samples.append({
                'timestamp': time.time(),
                'per_core': cpu_percents,
                'average': sum(cpu_percents) / len(cpu_percents)
            })
    
    def get_per_core_times(self) -> List[Dict]:
        if not self.cpu_times_start or not self.cpu_times_end:
            return []
        
        per_core_times = []
        for i, (start, end) in enumerate(zip(self.cpu_times_start, self.cpu_times_end)):
            on_cpu_time = (end.user + end.system) - (start.user + start.system)
            idle_time = end.idle - start.idle
            total_time = on_cpu_time + idle_time
            on_cpu_pct = (on_cpu_time / total_time * 100) if total_time > 0 else 0
            
            per_core_times.append({
                'core_id': i,
                'on_cpu_time': on_cpu_time,
                'idle_time': idle_time,
                'on_cpu_pct': on_cpu_pct,
                'off_cpu_pct': 100 - on_cpu_pct
            })
        return per_core_times
    
    def get_summary(self) -> dict:
        if not self.cpu_samples:
            return {'per_core_avg': [], 'overall_avg': 0.0, 'cores_used': 0, 'total_cores': 0}
        
        num_cores = len(self.cpu_samples[0]['per_core'])
        per_core_totals = [0.0] * num_cores
        
        for sample in self.cpu_samples:
            for i, val in enumerate(sample['per_core']):
                per_core_totals[i] += val
        
        num_samples = len(self.cpu_samples)
        per_core_avg = [total / num_samples for total in per_core_totals]
        overall_avg = sum(per_core_avg) / len(per_core_avg)
        consistently_active = [i for i, avg in enumerate(per_core_avg) if avg > 10.0]
        
        return {
            'per_core_avg': per_core_avg,
            'overall_avg': overall_avg,
            'num_samples': num_samples,
            'consistently_active_cores': consistently_active,
            'cores_used': len(consistently_active),
            'total_cores': num_cores,
            'per_core_times': self.get_per_core_times()
        }


class ThroughputMonitor:
    """Monitor and display throughput in real-time."""
    
    def __init__(self, stats: Stats, interval: float = 1.0):
        self.stats = stats
        self.interval = interval
        self.running = False
        self.thread = None
        self.samples = []
    
    def start(self):
        self.running = True
        self.samples = []
        self.thread = threading.Thread(target=self._monitor_loop, daemon=True)
        self.thread.start()
    
    def stop(self):
        self.running = False
        if self.thread:
            self.thread.join(timeout=2.0)
    
    def _monitor_loop(self):
        last_bytes = 0
        last_time = time.time()
        
        while self.running:
            time.sleep(self.interval)
            current_bytes = self.stats.bytes_read
            current_time = time.time()
            
            delta_bytes = current_bytes - last_bytes
            delta_time = current_time - last_time
            
            if delta_time > 0:
                throughput_mbps = (delta_bytes / (1024 * 1024)) / delta_time
                self.samples.append(throughput_mbps)
                
                total_read_gb = current_bytes / (1024 * 1024 * 1024)
                numbers = self.stats.numbers_read
                print(f"\r[Progress] Read: {total_read_gb:.2f} GB | "
                      f"Numbers: {numbers:,} | "
                      f"Throughput: {throughput_mbps:.2f} MB/s", 
                      end='', flush=True)
            
            last_bytes = current_bytes
            last_time = current_time
    
    def get_summary(self) -> dict:
        if not self.samples:
            return {'avg_throughput': 0.0, 'peak_throughput': 0.0, 'min_throughput': 0.0}
        return {
            'avg_throughput': sum(self.samples) / len(self.samples),
            'peak_throughput': max(self.samples),
            'min_throughput': min(self.samples)
        }


def find_max_single_thread(filepath: str, chunk_size: int = 64 * 1024 * 1024,
                           stats: Optional[Stats] = None) -> int:
    """Find maximum integer in text file using single thread."""
    max_val = float('-inf')
    buffer = ""
    
    with open(filepath, 'r') as f:
        while True:
            chunk = f.read(chunk_size)
            if not chunk:
                # Process remaining buffer
                if buffer:
                    for num_str in buffer.split(','):
                        num_str = num_str.strip()
                        if num_str:
                            val = int(num_str)
                            if val > max_val:
                                max_val = val
                            if stats:
                                stats.add_progress(0, 1)
                break
            
            if stats:
                stats.add_progress(len(chunk.encode('utf-8')), 0)
            
            # Combine with leftover from previous chunk
            buffer += chunk
            
            # Find last comma to avoid splitting a number
            last_comma = buffer.rfind(',')
            if last_comma == -1:
                continue  # No complete number yet
            
            # Process complete numbers
            complete_part = buffer[:last_comma]
            buffer = buffer[last_comma + 1:]
            
            for num_str in complete_part.split(','):
                num_str = num_str.strip()
                if num_str:
                    try:
                        val = int(num_str)
                        if val > max_val:
                            max_val = val
                        if stats:
                            stats.add_progress(0, 1)
                    except ValueError:
                        pass  # Skip invalid numbers
    
    return int(max_val) if max_val != float('-inf') else 0


def find_max_in_range(filepath: str, start: int, end: int, 
                      chunk_size: int, stats: Optional[Stats] = None) -> int:
    """Find maximum in a byte range of the file."""
    max_val = float('-inf')
    
    with open(filepath, 'r') as f:
        f.seek(start)
        
        # Skip partial number at start (unless at beginning)
        if start > 0:
            # Read until we find a comma
            while True:
                char = f.read(1)
                if not char or char == ',':
                    break
        
        remaining = end - f.tell()
        buffer = ""
        
        while remaining > 0:
            to_read = min(chunk_size, remaining)
            chunk = f.read(to_read)
            if not chunk:
                break
            
            remaining -= len(chunk)
            
            if stats:
                stats.add_progress(len(chunk.encode('utf-8')), 0)
            
            buffer += chunk
            
            # Find last comma
            last_comma = buffer.rfind(',')
            if last_comma == -1:
                continue
            
            complete_part = buffer[:last_comma]
            buffer = buffer[last_comma + 1:]
            
            for num_str in complete_part.split(','):
                num_str = num_str.strip()
                if num_str:
                    try:
                        val = int(num_str)
                        if val > max_val:
                            max_val = val
                        if stats:
                            stats.add_progress(0, 1)
                    except ValueError:
                        pass
        
        # Process remaining buffer
        if buffer:
            for num_str in buffer.split(','):
                num_str = num_str.strip()
                if num_str:
                    try:
                        val = int(num_str)
                        if val > max_val:
                            max_val = val
                        if stats:
                            stats.add_progress(0, 1)
                    except ValueError:
                        pass
    
    return int(max_val) if max_val != float('-inf') else 0


def find_max_multi_thread(filepath: str, num_threads: int, 
                          chunk_size: int = 64 * 1024 * 1024,
                          stats: Optional[Stats] = None) -> int:
    """Find maximum using multiple threads."""
    file_size = os.path.getsize(filepath)
    segment_size = file_size // num_threads
    
    segments = []
    for i in range(num_threads):
        start = i * segment_size
        end = file_size if i == num_threads - 1 else (i + 1) * segment_size
        segments.append((start, end))
    
    max_val = float('-inf')
    
    with ThreadPoolExecutor(max_workers=num_threads) as executor:
        futures = [
            executor.submit(find_max_in_range, filepath, start, end, chunk_size, stats)
            for start, end in segments
        ]
        
        for future in as_completed(futures):
            result = future.result()
            if result > max_val:
                max_val = result
    
    return int(max_val)


def print_tabulated_results(results: dict):
    """Print results in tabulated format."""
    print(f"\n{'='*80}")
    print(f"{'BENCHMARK RESULTS SUMMARY':^80}")
    print(f"{'='*80}")
    
    print(f"\n{'METRIC':<40} {'VALUE':>35}")
    print(f"{'-'*40} {'-'*35}")
    
    file_size_gb = results['file_size'] / (1024**3)
    print(f"{'File Size':<40} {file_size_gb:>32.2f} GB")
    print(f"{'Threads Used':<40} {results['num_threads']:>35}")
    print(f"{'Maximum Value Found':<40} {results['max_value']:>35}")
    print(f"{'Total Execution Time':<40} {results['elapsed_time']:>32.2f} sec")
    
    print(f"\n{'--- DISK THROUGHPUT ---':<40}")
    print(f"{'Average Throughput':<40} {results['throughput']['avg_throughput']:>30.2f} MB/s")
    print(f"{'Peak Throughput':<40} {results['throughput']['peak_throughput']:>30.2f} MB/s")
    
    print(f"\n{'--- I/O TIMING ---':<40}")
    io = results['io_stats']
    print(f"{'Wall Clock Time':<40} {io['wall_time']:>32.2f} sec")
    print(f"{'CPU Time (user+sys)':<40} {io['cpu_time']:>32.2f} sec")
    print(f"{'I/O Wait Time (estimated)':<40} {io['io_wait_time']:>32.2f} sec")
    print(f"{'I/O Wait Percentage':<40} {io['io_wait_pct']:>32.1f} %")
    
    print(f"\n{'--- CPU CORES ---':<40}")
    cpu = results['cpu']
    print(f"{'Total Logical Cores':<40} {cpu['total_cores']:>35}")
    print(f"{'Cores Actively Used':<40} {cpu['cores_used']:>35}")
    
    if cpu.get('per_core_times'):
        print(f"\n{'--- PER-CORE CPU TIME ---'}")
        print(f"{'Core':<8} {'On-CPU Time':>14} {'Idle Time':>14} {'On-CPU %':>12} {'Util %':>10}")
        print(f"{'-'*8} {'-'*14} {'-'*14} {'-'*12} {'-'*10}")
        
        for i, core_time in enumerate(cpu['per_core_times']):
            util_pct = cpu['per_core_avg'][i] if i < len(cpu['per_core_avg']) else 0
            marker = "*" if i in cpu.get('consistently_active_cores', []) else " "
            print(f"Core {i:<2}{marker} {core_time['on_cpu_time']:>13.2f}s "
                  f"{core_time['idle_time']:>13.2f}s {core_time['on_cpu_pct']:>11.1f}% {util_pct:>9.1f}%")
    
    print(f"\n{'='*80}\n")


def run_benchmark(filepath: str, num_threads: int = 1, chunk_size_mb: int = 64) -> dict:
    """Run the benchmark with full monitoring."""
    chunk_size = chunk_size_mb * 1024 * 1024
    file_size = os.path.getsize(filepath)
    
    print(f"\n{'='*60}")
    print(f"Text File Max Finder Benchmark (Python)")
    print(f"{'='*60}")
    print(f"File: {filepath}")
    print(f"File Size: {file_size / (1024**3):.2f} GB")
    print(f"Mode: {'Single-threaded' if num_threads == 1 else f'Multi-threaded ({num_threads} threads)'}")
    print(f"Format: Comma-separated integers (text)")
    print(f"{'='*60}\n")
    
    stats = Stats(start_time=time.time())
    cpu_monitor = CPUMonitor(interval=0.5)
    throughput_monitor = ThroughputMonitor(stats, interval=1.0)
    io_monitor = IOWaitMonitor()
    
    cpu_monitor.start()
    throughput_monitor.start()
    io_monitor.start()
    
    start_time = time.time()
    
    try:
        if num_threads == 1:
            max_val = find_max_single_thread(filepath, chunk_size, stats)
        else:
            max_val = find_max_multi_thread(filepath, num_threads, chunk_size, stats)
    finally:
        throughput_monitor.stop()
        cpu_monitor.stop()
    
    elapsed = time.time() - start_time
    
    return {
        'max_value': max_val,
        'elapsed_time': elapsed,
        'throughput': throughput_monitor.get_summary(),
        'cpu': cpu_monitor.get_summary(),
        'io_stats': io_monitor.get_stats(),
        'file_size': file_size,
        'num_threads': num_threads,
        'numbers_read': stats.numbers_read
    }


def main():
    parser = argparse.ArgumentParser(
        description='Find maximum integer in a comma-separated text file',
        epilog="""
Examples:
  python max_finder_text.py data.txt --threads 1
  python max_finder_text.py data.txt --threads 10
  python max_finder_text.py data.txt --threads 0  # all cores
        """
    )
    
    parser.add_argument('filepath', help='Path to text file')
    parser.add_argument('-t', '--threads', type=int, default=1,
                        help='Number of threads (0 = all cores)')
    parser.add_argument('-c', '--chunk-size', type=int, default=64,
                        help='Chunk size in MB')
    parser.add_argument('--csv', type=str, help='Append results to CSV')
    
    args = parser.parse_args()
    
    if not os.path.exists(args.filepath):
        print(f"Error: File not found: {args.filepath}", file=sys.stderr)
        sys.exit(1)
    
    num_threads = args.threads
    if num_threads == 0:
        num_threads = psutil.cpu_count(logical=True)
        print(f"Using all {num_threads} logical cores")
    
    try:
        results = run_benchmark(args.filepath, num_threads, args.chunk_size)
        print_tabulated_results(results)
        
        print(f"Summary: Found max={results['max_value']} in {results['elapsed_time']:.2f}s")
        
    except KeyboardInterrupt:
        print("\n\nBenchmark interrupted")
        sys.exit(130)


if __name__ == '__main__':
    main()

File: MaxFinderText.java
import java.io.*;
import java.lang.management.*;
import java.nio.*;
import java.nio.channels.*;
import java.nio.file.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.*;

/**
 * Text File Maximum Number Finder - Java Version
 * Finds the maximum integer in a comma-separated text file.
 * 
 * Supports:
 * - Single-threaded and multi-threaded modes
 * - Real-time throughput monitoring
 * - Per-thread CPU time tracking
 */
public class MaxFinderText {
    
    private static final AtomicLong bytesRead = new AtomicLong(0);
    private static final AtomicLong numbersRead = new AtomicLong(0);
    private static volatile boolean monitoringActive = false;
    
    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            printUsage();
            System.exit(1);
        }
        
        String filepath = args[0];
        int numThreads = 1;
        int chunkSizeMB = 64;
        String csvPath = null;
        
        // Parse arguments
        for (int i = 1; i < args.length; i++) {
            switch (args[i]) {
                case "--threads":
                    if (i + 1 < args.length) numThreads = Integer.parseInt(args[++i]);
                    break;
                case "--chunk-size":
                    if (i + 1 < args.length) chunkSizeMB = Integer.parseInt(args[++i]);
                    break;
                case "--csv":
                    if (i + 1 < args.length) csvPath = args[++i];
                    break;
            }
        }
        
        if (numThreads == 0) {
            numThreads = Runtime.getRuntime().availableProcessors();
            System.out.println("Using all " + numThreads + " logical cores");
        }
        
        File file = new File(filepath);
        if (!file.exists()) {
            System.err.println("Error: File not found: " + filepath);
            System.exit(1);
        }
        
        runBenchmark(filepath, numThreads, chunkSizeMB, csvPath);
    }
    
    private static void printUsage() {
        System.out.println("Usage: java MaxFinderText <filepath> [options]");
        System.out.println("\nOptions:");
        System.out.println("  --threads N      Number of threads (default: 1, 0 = all cores)");
        System.out.println("  --chunk-size MB  Read chunk size in MB (default: 64)");
        System.out.println("  --csv FILE       Append results to CSV file");
        System.out.println("\nExamples:");
        System.out.println("  java MaxFinderText data.txt --threads 1");
        System.out.println("  java MaxFinderText data.txt --threads 10");
        System.out.println("  java MaxFinderText data.txt --threads 0  # all cores");
    }
    
    private static void runBenchmark(String filepath, int numThreads, int chunkSizeMB,
                                      String csvPath) throws Exception {
        
        File file = new File(filepath);
        long fileSize = file.length();
        int chunkSize = chunkSizeMB * 1024 * 1024;
        int availableProcessors = Runtime.getRuntime().availableProcessors();
        
        System.out.println("\n" + "=".repeat(60));
        System.out.println("Text File Max Finder Benchmark (Java)");
        System.out.println("=".repeat(60));
        System.out.printf("File: %s%n", filepath);
        System.out.printf("File Size: %.2f GB%n", fileSize / (1024.0 * 1024.0 * 1024.0));
        System.out.printf("Mode: %s%n", numThreads == 1 ? "Single-threaded" : "Multi-threaded (" + numThreads + " threads)");
        System.out.printf("Format: Comma-separated integers (text)%n");
        System.out.printf("Available Processors: %d%n", availableProcessors);
        System.out.println("=".repeat(60) + "\n");
        
        // Reset counters
        bytesRead.set(0);
        numbersRead.set(0);
        monitoringActive = true;
        
        // Get thread MXBean for CPU time tracking
        ThreadMXBean threadMXBean = ManagementFactory.getThreadMXBean();
        
        // Throughput samples
        List<Double> throughputSamples = Collections.synchronizedList(new ArrayList<>());
        
        // Start monitor thread
        Thread monitorThread = new Thread(() -> {
            long lastBytes = 0;
            long lastTime = System.nanoTime();
            
            while (monitoringActive) {
                try {
                    Thread.sleep(1000);
                } catch (InterruptedException e) {
                    break;
                }
                
                long currentBytes = bytesRead.get();
                long currentNumbers = numbersRead.get();
                long currentTime = System.nanoTime();
                
                double deltaBytes = currentBytes - lastBytes;
                double deltaTime = (currentTime - lastTime) / 1e9;
                
                if (deltaTime > 0 && deltaBytes > 0) {
                    double speedMBps = (deltaBytes / (1024.0 * 1024.0)) / deltaTime;
                    throughputSamples.add(speedMBps);
                    
                    double totalGB = currentBytes / (1024.0 * 1024.0 * 1024.0);
                    System.out.printf("\r[Progress] Read: %.2f GB | Numbers: %,d | Throughput: %.0f MB/s",
                                     totalGB, currentNumbers, speedMBps);
                    System.out.flush();
                }
                
                lastBytes = currentBytes;
                lastTime = currentTime;
            }
        });
        monitorThread.setDaemon(true);
        monitorThread.start();
        
        // Record start times
        long wallStartTime = System.nanoTime();
        long cpuStartTime = threadMXBean.getCurrentThreadCpuTime();
        
        // Run the max finding
        long maxValue;
        Map<Long, Long> threadCpuTimes = new ConcurrentHashMap<>();
        
        if (numThreads == 1) {
            maxValue = findMaxSingleThread(filepath, chunkSize);
        } else {
            maxValue = findMaxMultiThread(filepath, numThreads, chunkSize, threadMXBean, threadCpuTimes);
        }
        
        // Record end times
        long wallEndTime = System.nanoTime();
        long cpuEndTime = threadMXBean.getCurrentThreadCpuTime();
        
        // Stop monitoring
        monitoringActive = false;
        monitorThread.interrupt();
        monitorThread.join(1000);
        
        // Calculate times
        double wallTime = (wallEndTime - wallStartTime) / 1e9;
        double cpuTime = (cpuEndTime - cpuStartTime) / 1e9;
        
        // For multi-threaded, sum thread CPU times
        if (!threadCpuTimes.isEmpty()) {
            double totalThreadCpu = 0;
            for (Long t : threadCpuTimes.values()) {
                totalThreadCpu += t / 1e9;
            }
            cpuTime = totalThreadCpu;
        }
        
        double ioWaitTime = Math.max(0, wallTime - cpuTime);
        double ioWaitPct = (ioWaitTime / wallTime) * 100;
        
        // Calculate throughput stats
        double avgThroughput = 0, peakThroughput = 0;
        if (!throughputSamples.isEmpty()) {
            for (double t : throughputSamples) {
                avgThroughput += t;
                peakThroughput = Math.max(peakThroughput, t);
            }
            avgThroughput /= throughputSamples.size();
        }
        double overallThroughput = (bytesRead.get() / (1024.0 * 1024.0)) / wallTime;
        
        // Print results
        System.out.println("\n\n" + "=".repeat(80));
        System.out.printf("%40s%n", "BENCHMARK RESULTS SUMMARY");
        System.out.println("=".repeat(80));
        
        System.out.printf("%n%-40s %35s%n", "METRIC", "VALUE");
        System.out.printf("%-40s %35s%n", "-".repeat(40), "-".repeat(35));
        
        System.out.printf("%-40s %32.2f GB%n", "File Size", fileSize / (1024.0 * 1024.0 * 1024.0));
        System.out.printf("%-40s %35d%n", "Threads Used", numThreads);
        System.out.printf("%-40s %35d%n", "Maximum Value Found", maxValue);
        System.out.printf("%-40s %,35d%n", "Numbers Processed", numbersRead.get());
        System.out.printf("%-40s %32.2f sec%n", "Total Execution Time", wallTime);
        
        System.out.printf("%n%-40s%n", "--- DISK THROUGHPUT ---");
        System.out.printf("%-40s %30.2f MB/s%n", "Average Throughput", avgThroughput > 0 ? avgThroughput : overallThroughput);
        System.out.printf("%-40s %30.2f MB/s%n", "Peak Throughput", peakThroughput);
        System.out.printf("%-40s %30.2f MB/s%n", "Overall Throughput", overallThroughput);
        
        System.out.printf("%n%-40s%n", "--- I/O TIMING ---");
        System.out.printf("%-40s %32.2f sec%n", "Wall Clock Time", wallTime);
        System.out.printf("%-40s %32.2f sec%n", "CPU Time", cpuTime);
        System.out.printf("%-40s %32.2f sec%n", "I/O Wait Time (estimated)", ioWaitTime);
        System.out.printf("%-40s %32.1f %%%n", "I/O Wait Percentage", ioWaitPct);
        
        System.out.printf("%n%-40s%n", "--- SYSTEM INFO ---");
        System.out.printf("%-40s %35d%n", "Available Processors", availableProcessors);
        System.out.printf("%-40s %35d%n", "Threads Used", numThreads);
        
        System.out.println("\n" + "=".repeat(80) + "\n");
        
        System.out.printf("Summary: Found max=%d in %.2fs @ %.0f MB/s avg throughput%n",
                         maxValue, wallTime, overallThroughput);
        
        // Save to CSV if requested
        if (csvPath != null) {
            saveResultsCsv(csvPath, fileSize, numThreads, maxValue, wallTime, 
                          avgThroughput > 0 ? avgThroughput : overallThroughput,
                          cpuTime, ioWaitTime, ioWaitPct);
            System.out.println("Results appended to: " + csvPath);
        }
    }
    
    private static long findMaxSingleThread(String filepath, int chunkSize) throws Exception {
        long maxVal = Long.MIN_VALUE;
        
        // Use smaller buffer to avoid memory issues
        int bufferSize = Math.min(chunkSize, 8 * 1024 * 1024);  // Max 8MB
        
        try (BufferedInputStream bis = new BufferedInputStream(
                new FileInputStream(filepath), bufferSize)) {
            
            StringBuilder currentNumber = new StringBuilder(32);
            byte[] byteBuffer = new byte[bufferSize];
            int bytesReadCount;
            
            while ((bytesReadCount = bis.read(byteBuffer)) != -1) {
                bytesRead.addAndGet(bytesReadCount);
                
                // Process byte by byte
                for (int i = 0; i < bytesReadCount; i++) {
                    byte b = byteBuffer[i];
                    
                    if (b == ',') {
                        // End of number
                        if (currentNumber.length() > 0) {
                            try {
                                long val = Long.parseLong(currentNumber.toString().trim());
                                if (val > maxVal) {
                                    maxVal = val;
                                }
                                numbersRead.incrementAndGet();
                            } catch (NumberFormatException e) {
                                // Skip invalid
                            }
                            currentNumber.setLength(0);
                        }
                    } else {
                        currentNumber.append((char) b);
                    }
                }
            }
            
            // Process last number
            if (currentNumber.length() > 0) {
                try {
                    long val = Long.parseLong(currentNumber.toString().trim());
                    if (val > maxVal) {
                        maxVal = val;
                    }
                    numbersRead.incrementAndGet();
                } catch (NumberFormatException e) {
                    // Skip invalid
                }
            }
        }
        
        return maxVal;
    }
    
    private static long findMaxMultiThread(String filepath, int numThreads, int chunkSize,
                                            ThreadMXBean threadMXBean,
                                            Map<Long, Long> threadCpuTimes) throws Exception {
        
        File file = new File(filepath);
        long fileSize = file.length();
        long segmentSize = fileSize / numThreads;
        
        ExecutorService executor = Executors.newFixedThreadPool(numThreads);
        List<Future<Long>> futures = new ArrayList<>();
        
        for (int i = 0; i < numThreads; i++) {
            final long start = i * segmentSize;
            final long end = (i == numThreads - 1) ? fileSize : (i + 1) * segmentSize;
            
            futures.add(executor.submit(() -> {
                long threadId = Thread.currentThread().getId();
                long cpuStart = threadMXBean.getThreadCpuTime(threadId);
                
                long result = findMaxInRange(filepath, start, end, chunkSize);
                
                long cpuEnd = threadMXBean.getThreadCpuTime(threadId);
                threadCpuTimes.put(threadId, cpuEnd - cpuStart);
                
                return result;
            }));
        }
        
        long maxVal = Long.MIN_VALUE;
        for (Future<Long> future : futures) {
            long result = future.get();
            if (result > maxVal) {
                maxVal = result;
            }
        }
        
        executor.shutdown();
        executor.awaitTermination(1, TimeUnit.HOURS);
        
        return maxVal;
    }
    
    private static long findMaxInRange(String filepath, long start, long end, int chunkSize) throws Exception {
        long maxVal = Long.MIN_VALUE;
        
        // Use smaller buffer to avoid memory issues
        int bufferSize = Math.min(chunkSize, 8 * 1024 * 1024);  // Max 8MB
        
        try (RandomAccessFile raf = new RandomAccessFile(filepath, "r")) {
            raf.seek(start);
            
            // Skip to next comma if not at start
            if (start > 0) {
                while (raf.getFilePointer() < end) {
                    int b = raf.read();
                    if (b == -1 || b == ',') break;
                }
            }
            
            StringBuilder currentNumber = new StringBuilder(32);  // For building current number
            byte[] byteBuffer = new byte[bufferSize];
            
            while (raf.getFilePointer() < end) {
                long remaining = end - raf.getFilePointer();
                int toRead = (int) Math.min(bufferSize, remaining);
                int bytesReadCount = raf.read(byteBuffer, 0, toRead);
                
                if (bytesReadCount <= 0) break;
                
                bytesRead.addAndGet(bytesReadCount);
                
                // Process byte by byte to avoid creating large strings
                for (int i = 0; i < bytesReadCount; i++) {
                    byte b = byteBuffer[i];
                    
                    if (b == ',') {
                        // End of number, parse it
                        if (currentNumber.length() > 0) {
                            try {
                                long val = Long.parseLong(currentNumber.toString().trim());
                                if (val > maxVal) {
                                    maxVal = val;
                                }
                                numbersRead.incrementAndGet();
                            } catch (NumberFormatException e) {
                                // Skip invalid
                            }
                            currentNumber.setLength(0);
                        }
                    } else {
                        currentNumber.append((char) b);
                    }
                }
            }
            
            // Process last number
            if (currentNumber.length() > 0) {
                try {
                    long val = Long.parseLong(currentNumber.toString().trim());
                    if (val > maxVal) {
                        maxVal = val;
                    }
                    numbersRead.incrementAndGet();
                } catch (NumberFormatException e) {
                    // Skip invalid
                }
            }
        }
        
        return maxVal;
    }
    
    private static void saveResultsCsv(String csvPath, long fileSize, int numThreads,
                                        long maxValue, double wallTime, double throughput,
                                        double cpuTime, double ioWaitTime, double ioWaitPct) throws Exception {
        
        boolean fileExists = new File(csvPath).exists();
        
        try (PrintWriter writer = new PrintWriter(new FileWriter(csvPath, true))) {
            if (!fileExists) {
                writer.println("timestamp,file_size_gb,num_threads,max_value,execution_time_sec," +
                              "avg_throughput_mbps,cpu_time_sec,io_wait_time_sec,io_wait_pct");
            }
            
            writer.printf("%s,%.4f,%d,%d,%.4f,%.2f,%.4f,%.4f,%.2f%n",
                         java.time.LocalDateTime.now().toString(),
                         fileSize / (1024.0 * 1024.0 * 1024.0),
                         numThreads,
                         maxValue,
                         wallTime,
                         throughput,
                         cpuTime,
                         ioWaitTime,
                         ioWaitPct);
        }
    }
}
